### PR TITLE
Improve min max performance and IEEE compliance

### PR DESCRIPTION
**Pull request by CoryCharlton**
## Description

- Reduce IsNan checks to improve performance
- Fix Min not returning NaN in all cases it should
- Implement Min(float) and Max(float)

## Motivation and Context

- Fixes nanoFramework/Home#1395
- Fixes nanoFramework/Home#1396

## How Has This Been Tested?

- On device
- Unit tests

## Types of changes

- [X] Improvement (non-breaking change that improves a feature, code or algorithm)
- [X] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [X] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [X] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [X] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by CoryCharlton** [comment]
Look like the build failure is because the [firmware changes](https://github.com/nanoframework/nf-interpreter/pull/2810) need to be merged first.

**Comment by CoryCharlton** [comment]
For reference:

### Before...

```
| MethodName | ItterationCount | Mean                  | Min   | Max   |
| -------------------------------------------------------------------- |
| Max_Double | 250             | 8.7599999999999998 ms | 0 ms  | 10 ms |
| Max_Float  | 250             | 43.52 ms              | 30 ms | 50 ms |
| Max_Int    | 250             | 22.199999999999998 ms | 20 ms | 30 ms |
| Min_Double | 250             | 13.32 ms              | 10 ms | 20 ms |
| Min_Float  | 250             | 28.959999999999999 ms | 20 ms | 40 ms |
| Min_Int    | 250             | 13.72 ms              | 10 ms | 20 ms |
```

### After...

```
| MethodName | ItterationCount | Mean     | Min   | Max   |
| ------------------------------------------------------- |
| Max_Double | 250             | 6.48 ms  | 0 ms  | 10 ms |
| Max_Float  | 250             | 7.04 ms  | 0 ms  | 10 ms |
| Max_Int    | 250             | 7.48 ms  | 0 ms  | 10 ms |
| Min_Double | 250             | 7 ms     | 0 ms  | 10 ms |
| Min_Float  | 250             | 7.8 ms   | 0 ms  | 10 ms |
| Min_Int    | 250             | 10.44 ms | 10 ms | 20 ms |
```

**Comment by josesimoes** [comment]
> Look like the build failure is because the [firmware changes](https://github.com/nanoframework/nf-interpreter/pull/2810) need to be merged first.

>> That is correct. Currently a PR on another repo that depends on nanoCLR library has to have that one merged before it can be consumed in the pipeline.

File: nanoFramework.System.Math.Benchmark/Program.cs [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
         {
-            BenchmarkRunner.Run(typeof(IAssemblyHandler).Assembly);
+#if DEBUG
+            Console.WriteLine("Benchmarks should be run in a release build.");
+            Debugger.Break();
+            return;
+#endif
+            Console.WriteLine("Running benchmarks...");
+
+            //BenchmarkRunner.Run(typeof(IAssemblyHandler).Assembly);
```
Comment (josesimoes) [inline nanoFramework.System.Math.Benchmark/Program.cs:19]: Please remove commented code, if not needed.

File: nanoFramework.System.Math/Math.cs [hunk 2]
```diff
@@ -353,7 +353,8 @@ public static int Abs(int value)
         /// <returns>Parameter <paramref name="val1"/> or <paramref name="val2"/>, whichever is larger.</returns>
         public static int Max(int val1, int val2)
         {
-            return MathInternal.Max(val1, val2);
+            // TODO: Copy this over to MathInternal and remove the native version
```
Comment (josesimoes) [inline nanoFramework.System.Math/Math.cs:356]: Yes, that's the way to go. mscorlib is always present and this will be available everywhere.

File: nanoFramework.System.Math/Math.cs [hunk 3]
```diff
@@ -386,7 +385,8 @@ public static float Max(float val1, float val2)
         /// <returns>Parameter <paramref name="val1"/> or <paramref name="val2"/>, whichever is smaller.</returns>
         public static int Min(int val1, int val2)
         {
-            return MathInternal.Min(val1, val2);
+            // TODO: Copy this over to MathInternal and remove the native version
```
Comment (josesimoes) [inline nanoFramework.System.Math/Math.cs:388]: Same here: move to mscorlib.

**Review by josesimoes — COMMENTED** [review]
Impressive! Thanks for all the work on this. 
Great usage of the benchmark library! 😉 

Overall all good. We're better moving the Max and  Min to mscorlib. 
Even if that's not release rigth away. It won't break anything.

File: nanoFramework.System.Math/Math.cs (hunk 2 above)
Comment (CoryCharlton) [reply nanoFramework.System.Math/Math.cs:356]: Since there is a cost incurred with a stack jump (even if it's a very minor one) do you have any objection to me leaving this implementation here and submitting a separate PR for the change in `MathInternal`.

This is a case of extreme micro optimization but my thoughts are that "math" is core so if there is ever a valid use case for over optimizing this is it :D

Comment (josesimoes) [reply nanoFramework.System.Math/Math.cs:356]: 😅 I can't think of anything against that. In lack of a different opinion, just go for it.

Comment (CoryCharlton) [reply nanoFramework.System.Math/Math.cs:356]: - https://github.com/nanoframework/nf-interpreter/pull/2815
- https://github.com/nanoframework/CoreLibrary/pull/208

Comment (josesimoes) [inline nanoFramework.System.Math/Math.cs:356]:
```suggestion
```

File: nanoFramework.System.Math/Math.cs (hunk 3 above)
Comment (josesimoes) [inline nanoFramework.System.Math/Math.cs:388]:
```suggestion
```

File: Tests/MathUnitTests/Max_Tests.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using nanoFramework.TestFramework;
+
+namespace MathUnitTests
+{
+    [TestClass]
+    public class Max_Tests
+    {
+        [TestMethod]
+        public void Max_Double_returns_greater_value_INVESTIGATE()
```
Comment (josesimoes) [inline Tests/MathUnitTests/Max_Tests.cs:10]: @CoryCharlton OK to "skip" this test (and the one in Min class) for the time being and then get back to it after everything is updated?

Comment (CoryCharlton) [reply Tests/MathUnitTests/Max_Tests.cs:10]: That's my thoughts. The current tests were implemented to assume that a passing test was +/- 0.0001:

`Assert.IsFalse((expected[i] - actual) > 0.0001d || (expected[i] - actual) < -0.0001d)`

So the failure of the `AreEqual` calls here are the existing behavior and this is not a regression.

I did not know enough how the system is working to understand if this was a technical limitation or an actual defect. However my initial thought was that if I pass in 0.12345 then the return value should be exactly the same but the precision is changing shortly.

Long story short I'll remove these tests for now and open a defect (and investigate it) later 👍

Comment (CoryCharlton) [reply Tests/MathUnitTests/Max_Tests.cs]: I opened a defect for this here: https://github.com/nanoframework/Home/issues/1401
